R/add-subvariable.R

```diff
@@ -12,25 +13,55 @@
 #' }
 #' @export
 addSubvariable <- function (variable, subvariable) {
-    ## Get subvariable URL or URLs, depending on how many supplied
-    new.urls <- urls(subvariable)
+    subvariable <- add_subvar_def(variable, subvariable)
 
-    ## Store these for post workaround
-    subvar.urls <- subvariables(tuple(variable))
+    ## There is some inconsistency in the output order after patching the subvariable catalog
+    ## this stores the proper order.
+    order <- c(names(subvariables(variable)), vapply(subvariable, name, ""))
 
+    ## Get subvariable URL or URLs, depending on how many supplied
+    new.urls <- urls(subvariable)
     ## Do the adding
     crPATCH(shojiURL(variable, "catalogs", "subvariables"),
         body=toJSON(sapply(new.urls, emptyObject, simplify=FALSE)))
-
-    ## Workaround because apparently bind/rebind isn't retaining the order
-    crPATCH(self(variable),
-        body=toJSON(list(subvariables=I(c(subvar.urls, new.urls)))))
-
+    variable <- refresh(variable)
+    subvariables(variable) <- subvariables(variable)[order]
     ## Refresh and return
     dropCache(datasetReference(variable))
-    invisible(refresh(variable))
+    return(invisible(refresh(variable)))
 }
 
 #' @rdname addSubvariable
 #' @export
 addSubvariables <- addSubvariable
+
+add_subvar_def <- function(var, subvar){
+    ## Input can be a variable, subvariable, dataset subset or
+    ## a mixed or uniform list of variables and subvariables this
+    ## wraps single entries in a list for type consistency.
+    if (inherits(subvar, "VariableDefinition") ||
+            is.variable(subvar)) {
+        ## wrap single variables in list
+        subvar <- list(subvar)
+    }
+
+    vardefs <- vapply(subvar,
+        function(x) inherits(x, "VariableDefinition"),
+        logical(1))
+
+    if (any(vardefs)) {
+        ds <- loadDataset(datasetReference(var))
+        ds <- addVariables(ds, subvar[vardefs])
```

Comment:
A more efficient way to do this is to call POSTNewVariable directly, what addVariables does internally. What you really want out of all of this is the URL(s) of the variables being added--that's what you're going to PATCH with in addSubvariables--and that's what POSTNewVariable returns. That way, you don't have to refresh the dataset and try to find the variable entity in the dataset. Then this function should return the vector of urls.

Reply:
Ah great, that makes a lot more sense